drivers/scheduler/k8s/k8s.go

```diff
@@ -2237,6 +2238,95 @@ func (k *K8s) WaitForDestroy(ctx *scheduler.Context, timeout time.Duration) erro
+// provided in the exclude list
+func (k *K8s) SelectiveWaitForTermination(ctx *scheduler.Context, timeout time.Duration, excludeList []node.Node) error {
+	t := func() (interface{}, bool, error) {
+		podNames, err := filterPodsByNodes(ctx, excludeList)
+		if err != nil {
+			return nil, true, err
+		}
+		if len(podNames) > 0 {
+			return nil, true, fmt.Errorf("pods %s still present in the system", podNames)
+		}
+		return nil, false, nil
+	}
+
+	if _, err := task.DoRetryWithTimeout(t, timeout, DefaultRetryInterval); err != nil {
+		return err
+	}
+	return nil
+}
+
+// filterPodsByNodes returns a list of pod names started as part of the provided context
+// and not running on the nodes provided in the exclude list
+func filterPodsByNodes(ctx *scheduler.Context, excludeList []node.Node) ([]string, error) {
+	allPods := make(map[types.UID]corev1.Pod)
+	excludedNodePods := make(map[types.UID]corev1.Pod)
+	namespaces := make(map[string]string)
+	for _, specObj := range ctx.App.SpecList {
+		var pods []corev1.Pod
+		var err error
+		if obj, ok := specObj.(*appsapi.Deployment); ok {
+			if pods, err = k8sApps.GetDeploymentPods(obj); err != nil {
+				return nil, &scheduler.ErrFailedToValidateAppDestroy{
+					App:   ctx.App,
+					Cause: fmt.Sprintf("Failed to get deployment: %v. Err: %v", obj.Name, err),
+				}
+			}
+			namespaces[obj.Namespace] = ""
+
+		} else if obj, ok := specObj.(*appsapi.StatefulSet); ok {
+			if pods, err = k8sApps.GetStatefulSetPods(obj); err != nil {
+				return nil, &scheduler.ErrFailedToValidateAppDestroy{
+					App:   ctx.App,
+					Cause: fmt.Sprintf("Failed to get statefulset: %v. Err: %v", obj.Name, err),
+				}
+			}
+			namespaces[obj.Namespace] = ""
+
+		} else if obj, ok := specObj.(*corev1.Pod); ok {
+			pod, err := k8sCore.GetPodByUID(obj.UID, obj.Namespace)
+			if err != nil && err != schederrors.ErrPodsNotFound {
+				return nil, &scheduler.ErrFailedToValidatePodDestroy{
+					App:   ctx.App,
+					Cause: fmt.Sprintf("Failed to get pod: %v. Err: %v", obj.Name, err),
+				}
+			}
+			pods = []corev1.Pod{*pod}
+			namespaces[obj.Namespace] = ""
+		}
+		for _, pod := range pods {
+			allPods[pod.UID] = pod
+		}
+	}
+
+	for _, excludedNode := range excludeList {
+		for _, namespace := range namespaces {
+			podList, err := k8sCore.GetPodsByNode(excludedNode.Name, namespace)
+			if err != nil {
+				return nil, &scheduler.ErrFailedToValidateAppDestroy{
+					App:   ctx.App,
+					Cause: fmt.Sprintf("Failed to get pods by node %v. Err: %v", excludedNode.Name, err),
+				}
+			}
+			for _, pod := range podList.Items {
+				excludedNodePods[pod.UID] = pod
+			}
+		}
+	}
+
+	// Compare the two pod maps. Create a list of pod names which are not running
+	// on the excluded nodes
+	var podNames []string
+	for _, pod := range allPods {
+		if _, exists := excludedNodePods[pod.UID]; !exists {
+			podNames = append(podNames, pod.Name)
```

Comment:
nit: Probably fine in practice but it seems odd that here we are returning the pod names without namespace presumably under the assumption that all pods are in the same namespace. But the code above tries to handle multiple namespaces.

Reply:
I was also thinking of adding node names as well to indicate where the pods are running. Will do.